From 8ef595f9fd8808a7d210a63e50173e67233f7018 Mon Sep 17 00:00:00 2001
From: Ivan Vecera <ivecera@redhat.com>
Date: Mon, 8 Sep 2014 16:14:22 -0400
Subject: [ethernet] be2net: use adapter->flags to track SRIOV state

Message-id: <1410192881-30702-60-git-send-email-ivecera@redhat.com>
Patchwork-id: 92441
O-Subject: [RHEL7.1 PATCH 59/78] be2net: use adapter->flags to track SRIOV state
Bugzilla: 1088575
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>

BZ: #1088575

Upstream commit(s):
commit f174c7ec10d99e62f7722d0608b2a881ef091d21
Author: Vasundhara Volam <vasundhara.volam@emulex.com>
Date:   Thu Jul 17 16:20:31 2014 +0530

    be2net: use adapter->flags to track SRIOV state

    The driver so far used adapter->num_vfs value to check if SR-IOV is enabled
    or not. But, the patch bec84e6("create optimal number of queues on SR-IOV
    config") changed this logic. The adapter->num_vfs value is validated and set
    much before SR-IOV is enabled. So, we now use an explicit flag to track
    SR-IOV enabled state.

    Signed-off-by: Vasundhara Volam <vasundhara.volam@emulex.com>
    Signed-off-by: Sathya Perla <sathya.perla@emulex.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Ivan Vecera <ivecera@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/emulex/benet/be.h b/drivers/net/ethernet/emulex/benet/be.h
index 2025dd0..d8a2f66 100644
--- a/drivers/net/ethernet/emulex/benet/be.h
+++ b/drivers/net/ethernet/emulex/benet/be.h
@@ -372,6 +372,7 @@ enum vf_state {
 };
 
 #define BE_FLAGS_LINK_STATUS_INIT  1
+#define BE_FLAGS_SRIOV_ENABLED   (1 << 2)
 #define BE_FLAGS_WORKER_SCHEDULED  (1 << 3)
 #define BE_FLAGS_VLAN_PROMISC   (1 << 4)
 #define BE_FLAGS_MCAST_PROMISC   (1 << 5)
@@ -525,7 +526,8 @@ struct be_adapter {
 
 #define be_physfn(adapter)  (!adapter->virtfn)
 #define be_virtfn(adapter)  (adapter->virtfn)
-#define sriov_enabled(adapter)  (adapter->num_vfs > 0)
+#define sriov_enabled(adapter)  (adapter->flags & \
+      BE_FLAGS_SRIOV_ENABLED)
 
 #define for_all_vfs(adapter, vf_cfg, i)     \
  for (i = 0, vf_cfg = &adapter->vf_cfg[i]; i < adapter->num_vfs; \
diff --git a/drivers/net/ethernet/emulex/benet/be_main.c b/drivers/net/ethernet/emulex/benet/be_main.c
index ca1e6b9..fe696d5 100644
--- a/drivers/net/ethernet/emulex/benet/be_main.c
+++ b/drivers/net/ethernet/emulex/benet/be_main.c
@@ -3045,6 +3045,7 @@ static void be_vf_clear(struct be_adapter *adapter)
 done:
  kfree(adapter->vf_cfg);
  adapter->num_vfs = 0;
+ adapter->flags &= ~BE_FLAGS_SRIOV_ENABLED;
 }
 
 static void be_clear_queues(struct be_adapter *adapter)
@@ -3237,6 +3238,8 @@ static int be_vf_setup(struct be_adapter *adapter)
    goto err;
   }
  }
+
+ adapter->flags |= BE_FLAGS_SRIOV_ENABLED;
  return 0;
 err:
  dev_err(dev, "VF setup failed\n");
-- 
1.7.1